### ISSUE TITLE
Add support for background location tracking

### ISSUE THREAD

**Issue by martijn00** [BaseflowIT/flutter-geolocator#53]
Also see: https://github.com/flutter/plugins/pull/656

**Comment by eliashdezr** [comment]
This is one of the main reasons I haven't switch one of my apps to Flutter, background location tracking is a must.

**Comment by aidandavis** [comment]
I'm also looking into this. Wanting to build an app in a corporate setting with a background GPS tracking requirement. flutter/plugins#656 (mentioned above) isn't on Pub yet and only supports iOS.
I'll subscribe to this thread for any updates.

**Comment by aidandavis** [comment]
I ended up writing my own implementation, based off [this Google sample](https://github.com/googlesamples/android-play-location/tree/master/LocationUpdatesForegroundService) that runs when the app closes.

I'm pushing the location data to firebase, so I didn't need to send anything back up the method channel. I just upload straight to Firebase using native code when a location result arrives. This means I can't generalise it to help this plugin out, but that sample repo might be useful.

**Comment by velegip** [comment]
I found a solution to generalize your @aidandavis method.
In this [packege](https://github.com/talkdirty/background_location_updates) all locations stored in sqlite.
If your foreground service can accept store provider sqlite/firebase/... .
With MethodChannel we can get the db file location, and the flutter app can read it.

**Comment by aidandavis** [comment]
That's definitely a way to do it. The problem still remaining is doing something with the data while the app is in background or the sytem chooses to close it.

With regards to Doze, it should only apply when the phone is stationary (like on a table). There's no point tracking location updates at that point anyway - the last location sent prior to Doze would still be accurate.

**Comment by sestegra** [comment]
On top of this, Dart code could be embedded as the same way implemented in alarm_manager_android. The Android callback calls Dart code without the app running.

**Comment by dark-chocolate** [comment]
Guys, any update on this? This feature can add wings to this plugin. It is must needed.

**Comment by ibrierley** [comment]
It's maybe also worth noting that some devices can't get gps locations with the screen off. I've been stuck on this one for a while, and finally found it's down to the devices chipset. So it may be worth double checking that as well, especially if it's a low battery device like a watch.

**Comment by dark-chocolate** [comment]
@ibrierley I am not sure about the wearables but all the devices running API >= 16 are capable of getting location updates even when the screen is off. I tested that myself too.

**Comment by ibrierley** [comment]
Yes, I've mainly been testing on a Huawei watch 2, and read a thread where there was a post with I think a Huawei engineer (may be wrong), saying it was set in the chipset, so they couldn't fix it (I'll post here if I can find the thread, but haven't seen it on a quick search). 

I've ended up turning the screen black, but I hate that solution :) (but it works, just not quite sure if it saves that much battery, so may end up forcing the screen on with something displaying).

**Comment by dark-chocolate** [comment]
Here how I got that working in iOS. 
https://github.com/BaseflowIT/flutter-geolocator/issues/150

**Comment by angel1st** [comment]
@martijn00  I wonder if [Geofence plugin](https://medium.com/flutter-io/executing-dart-in-the-background-with-flutter-plugins-and-geofencing-2b3e40a1a124) won't give a boost on this feature. I believe background execution would be a crucial one and many of us are expecting it. Let me know what do you think...

**Comment by ashihCT** [comment]
Hope you guys can add BG service soon :)

**Comment by kocholes** [comment]
Hi!
What is the current status of this issue? is someone working on this feature? or has been postponed?

**Comment by lambasoft** [comment]
Hey guys what are your thoughts on https://pub.dartlang.org/packages/flutter_background_geolocation 
Doesn't this plugin provide what is asked here?

**Comment by kostiaalyo** [comment]
> Hey guys what are your thoughts on https://pub.dartlang.org/packages/flutter_background_geolocation
> Doesn't this plugin provide what is asked here?

It does, but it costs $300+

**Comment by christocracy** [comment]
@kostiaalyo It’s > 5 years old with thousands of paying customers and supported full time.  Flutter is merely the latest port.

**Comment by akinsete** [comment]
See this article
https://medium.com/flutter-io/executing-dart-in-the-background-with-flutter-plugins-and-geofencing-2b3e40a1a124

**Comment by gregology** [comment]
@christocracy, I think your package is worth every cent however I'm just too poor atm :) Watching this space

**Comment by christocracy** [comment]
@gregology I see you used to work for Shopify.  Contact me for a free key.

**Comment by vishweshsoni** [comment]
hi there is it possible to track live location @christocracy   freely?

**Comment by dark-chocolate** [comment]
@vishweshsoni Please consider asking other plugins doubt on their respective thread. Thank you

**Comment by nikoskip** [comment]
Looking forward to this!

**Comment by mateuszmatyszkowicz** [comment]
Do you think it will be possible to start work on in it in this year?

**Comment by audkar** [comment]
> Do you think it will be possible to start work on in it in this year?

Anyone can start working immediately. Contributions are welcome. 

Maintainers of this project does quite good job on accepting accepting incoming changes from community.

**Comment by mateuszmatyszkowicz** [comment]
Ok, gimmie resources, and place in code where to start and I try to do my best :D I could spend few weeks to solve it but firstly I need to get in to android deeper :(  , I'm just curios if someone is working on it ;)

**Comment by audkar** [comment]
I appreciate your enthusiasm :1st_place_medal: 

> gimmie resources, and place in code where to start

I believe that you will have to touch and understand this whole plugin.

---------

However... Few high level design problems should considered before you start coding.
1. On android newer Android (API 26) only way to get high precision location updates is using Foreground service. Latest (Android Q) also will have additional permission (for getting location in background). So need to address 3 different behavior to have compatibility for: android pre 26, android 26 and android Q
2. To use Foreground service you need to show system notification. Probably every application will want to have customized notification.
3. Compatibility with iOS. Geolocator API should hide differences between platform as much as possible.
4. Foreground Service can be stopped & restarted same as Activity. Need logic to handle this. Flutter doesn't have easy way to do this. Only example I know is `alarm_manager` plugin. But it buggy as well and doesn't handle all cases. I don't know any open source Flutter library which does this correctly.

------

For our case we need to track more data for long periods of time (location, bluetooth, sensors). So making every plugin handle long running logic separately is not suitable for use because every plugin would have separate Foreground service and user would see three different Notifications.

**Comment by lambasoft** [comment]
I hope google would add this feature in flutter out of the box :p

**Comment by nsmith1024** [comment]
Hello, is this issue closed? I need the background geolocation too for android and ios for flutter! Which is the best lib for this?

**Comment by moovida** [comment]
Just to add my +1 to this feature request. For some applications this is a showstopper. 
Would be amazing to have it.

**Comment by Martibis** [comment]
+1

**Comment by Kelvinkins** [comment]
This is highly needed in my project... I was impressed by the one I found on pub,  but 300USD?

**Comment by shah-xad** [comment]
I was also facing this issue and I developed my own plugin for background location updates, here it is, [flutter_background_location](https://pub.dev/packages/flutter_background_location) , but its for android only as I don't own a mac..

**Comment by milesgranger** [comment]
+1 Also looking forward to this.

**Comment by Zeissus** [comment]
+1 This is very much needed!

**Comment by shishirraven** [comment]
+ This feature is needed. Please add this!

**Comment by shishirraven** [comment]
+ 1

**Comment by dark-chocolate** [comment]
@milesgranger @Zeissus  @shishirraven and others

If you have this issue, instead of writing `+1` in comment, give a thumbs up to the issue, it makes more sense.

**Comment by AlexCampos95** [comment]
any one saw this guy?

https://pub.dev/packages/flutter_background_geolocation

it works?

**Comment by AlexCampos95** [comment]
Does anyone know if this plugin works?

[flutter_background_geolocation](https://pub.dev/packages/flutter_background_geolocation)

I couldn't use it ... and I don't even know if it's free

**Comment by christocracy** [comment]
@AlexCampos95 Why don't you ask me at *my* repo?

**Comment by lyseiha** [comment]
Any solution yet?

**Comment by Slaheddine** [comment]
any solution ?

**Comment by fhamborg** [comment]
any?

**Comment by akinsete** [comment]
Don't know if this is helpful, was working on this some months ago.
https://github.com/akinsete/Flutter-Background-Location-Service

Try it out.

**Comment by estevez-dev** [comment]
The only thing that really works for me is [workmanager](https://pub.dev/packages/workmanager) + [geolocator](https://pub.dev/packages/geolocator). If battery saving will be disabled for your app, this will work constantly even after phone restart.

**Comment by akinsete** [comment]
Hi, Yegor Vialov

This seems like a good idea. Do you have a sample project that implement
both for background location monitoring? If you do can you please share.

Thanks

**Comment by estevez-dev** [comment]
@akinsete , HA Client: https://github.com/estevez-dev/ha_client

**Comment by MazEbeid** [comment]
Workmanager is inconsistent in iOS you dont have any control over when a task would run so it wont work for realtime location updates purposes. On Android the minimum frequency is 15mins so this doesnt work as well.

We really need this to be implemented in flutter out of the box.

**Comment by ZM123** [comment]
I don't have a lot of experience with Android, but it looks like there's a [separate call](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient#requestLocationUpdates(com.google.android.gms.location.LocationRequest,%20android.app.PendingIntent)) for background location updates compared to foreground ones (this is the one that is currently used in this plugin). Not sure how it would work but I imagine we would want to somehow use the optimal function based on which state the app is in.

Are there any other updates on this? This is a pretty key feature for a location tracking package and would be really nice to have if this is to be the number 1 location package for flutter

**Comment by jamesdixon** [comment]
@christocracy 's plugin is well worth the money and he responds very, very quickly. If background location is important to your application, it's a nominal cost.

**Comment by badaween-ahmed** [comment]
@jamesdixon that library is great. But I have a task manager app where tasks have location notification. Using geolocator + flutter local notification I can trigger push notification when the app is foreground but doesn't work when the app goes background. flutter_background_geolocation is too much for my requirement.

**Comment by steinmetz** [comment]
Maybe this lib can help you:
https://github.com/rekab-app/background_locator

**Comment by RomanJos** [comment]
> 
> 
> Maybe this lib can help you:
> https://github.com/rekab-app/background_locator

Maaaaan you find a gem here its EXACTLY what I need the only problem is that it lack documentation

**Comment by steinmetz** [comment]
> > Maybe this lib can help you:
> > https://github.com/rekab-app/background_locator
> 
> Maaaaan you find a gem here its EXACTLY what I need the only problem is that it lack documentation

I managed to execute their example... It works ok, sometimes it seems that it stops recording the coordinates when I close the main app and it executes the service in background.. But I have to test it again...

Let me know if you need help to execute it :)

**Comment by RomanJos** [comment]
It work well for me the only problem is to save on disk in the callback function, the service don't register the plugins native code and thus I can't make native calls using path_provider for example, also if the app is killed the Isolate will send to the ReceivePort from the killed FlutterActivity so nothing is listening. On restart I need to kill the isolate and then recreate it using a new port but I don't think its the best solution lol.
I also found [this example](https://github.com/ppicas/flutter-android-background) that is quite similar to what the library do (have a service that execute dart code with a notification) but they register the plugins entry in the Kotlin file so they are able to store things on the disk
If you know how I can retrieve the LatLng from the callback function it would be awesome🙏

**Comment by Ashar2shahid** [comment]
@christocracy would appreciate if you can send a key this way because I have tried all the solutions above and none seem to work consistently (either they terminate after app closes or stop recording). I am currently building a corona tracker at the request of the director of my local hospital. If you can help me out it would be appreciated.

**Comment by christocracy** [comment]
@Ashar2shahid Please contact me through my web-site https://www.transistorsoft.com

**Comment by RomanJos** [comment]
> I am currently building a corona tracker at the request of the director of my local hospital. If you can help me out it would be appreciated.

You probably want to fork some existing project on Github instead of starting from scratch

**Comment by christocracy** [comment]
@RomanJos @Ashar2shahid https://corona-trace.github.io/

**Comment by tatiblockchain** [comment]
@christocracy This is exactly what i was trying to build - searching for background-location-plugins for flutter led me here. However the link on this website: https://corona-trace.github.io/
for Slack channel is not working. Do you have another link for slack?

I have gone through the repo - this will save me a lot of work, when i find the $300 plus. Apple is trying to build something similar using bluetooth - they are more concerned about privacy - but i think - location privacy is a small price to pay in these times, building it like this, using google takeout etc is exactly what i had in mind.

**Comment by christocracy** [comment]
@tatiblockchain please come over to my repo to discuss, please.

**Comment by MazEbeid** [comment]
Why is this such a big deal to have out of the box?

**Comment by mvanbeusekom** [comment]
@MazEbeid what do you mean exactly? 

I would love to be able to support it, I just don't have the knowledge required to build it. Also didn't find the time to learn how to do it. Please feel free to create a pull-request or maybe you can provide some good resource on how to do it so I can build it (aside from the tutorial from the [Flutter site](https://flutter.dev/docs/development/packages-and-plugins/background-processes), I have read this many times but having a hard time getting a solid understanding)?

**Comment by MazEbeid** [comment]
I meant why is the Flutter team not adding a nice and maintainable plugin for background location fetching.

I have used this plugin [background_locator](https://pub.dev/packages/background_locator) which works well but at the moment it doesn't support location updates on iOS when the app is terminated

**Comment by TayfunCesur** [comment]
Joined the club without any hope =(

**Comment by TheMisir** [comment]
I'm currently using `background_locator` package. It's not perfect but works

**Comment by Flavsditz** [comment]
`background_locator` is nice... but it has its problems.

I am very interested on finding something more robust (hopefully this here) and I would not mind on having some **best practices** for storing these values while for example the app is in the background or the screen is off.

I have found that a lot of the options out there are not Isolate-Safe and cause problems such as corrupted data or even just access errors which is not ideal if dealing with a background solution.

**Comment by TheMisir** [comment]
@Flavsditz I'm currently planning developing native plugin for bg tracking. I'm probably going to track location by displaying chat bubble (I might be wrong but I think android requires foreground service to keep service alive) on android. But I don't have any idea about tracking on iOS.

**Comment by Itmaxglobal** [comment]
I would love to use this plugin, but due to the unavailability of the background location updates, I am forced to choose something else that can provide this feature even if I have to pay $300+. This feature is very important for any application with geolocation.

**Comment by MazEbeid** [comment]
> I would love to use this plugin, but due to the unavailability of the background location updates, I am forced to choose something else that can provide this feature even if I have to pay $300+. This feature is very important for any application with geolocation.

Have you tried this plugin? It has been recently updated and it should support background location.
https://github.com/rekab-app/background_locator

**Comment by ndhbr** [comment]
@mvanbeusekom Do you have intentions to include this? It would be incredibly cool. A lot of people need that.

**Comment by mvanbeusekom** [comment]
@ndhbr yes I do have the intention to support fetching position updates in the background. However I don't really find the time to do so as of yet. Every time I get started something else (more important) comes up. Now for example iOS 14 is released which introduces a new API for location permissions which needs to be supported for users to be able to use the geolocator on iOS 14.

So if you feel comfortable it would be great if you can make a start, maybe rough edges. The idea I am having is to provide developers an option to register a Dart callback method which will be executed (only) when the App is in the background each time a position update arrives (as described in this [article](https://medium.com/flutter/executing-dart-in-the-background-with-flutter-plugins-and-geofencing-2b3e40a1a124)). It would already help a lot if we can make a plan. Break the work that needs to be done into smaller pieces.

**Comment by ndhbr** [comment]
@mvanbeusekom 
Ah okay, thanks for the update. I can understand that. Thank you for your efforts.

I already know the article almost by heart, but it's still not easy.

I've been trying for a week to get background_locator to run stable (often doesn't work on Android) and they work with isolates. An isolate starts and gets the location in the background. My current procedure is to fill a file byte by byte in the callback and read the file from the last byte in the UI. Unfortunately Android writes the location in the callback only every few minutes and so I lose almost all locations. Under iOS the procedure works well.

Maybe this is also some inspiration.

**Comment by bhanuka96** [comment]
Is there any **STABLE** alternative way to handle background location?

And Are you guys decided to add this feature in the future?

**Comment by TheMisir** [comment]
> Is there any **STABLE** alternative way to handle background location?
> 
> And Are you guys decided to add this feature in the future?

https://github.com/rekab-app/background_locator works well.

**Comment by bhanuka96** [comment]
@TheMisir Thank you. Are you sure is it working well and is it okay for the production app? There are crash reports too in here https://github.com/rekab-app/background_locator/issues

**Comment by TheMisir** [comment]
> @TheMisir Thank you. Are you sure is it working well and is it okay for the production app? There are crash reports too in here https://github.com/rekab-app/background_locator/issues

I had no choice. So we're using it in production. Yes it's not works perfectly but it works if configured correctly.  (Sometimes issues on our projects was caused by wrong configuration).

**Comment by ndhbr** [comment]
> Is there any **STABLE** alternative way to handle background location?
> 
> And Are you guys decided to add this feature in the future?

Yes, because I urgently needed the function in production I bought a license for https://www.transistorsoft.com/shop/products/flutter-background-geolocation. I have been using it for 1 month in production and it works fine so far.

**Comment by Dereumaux-Adrien** [comment]
Once I add:
`    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
    <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
    <uses-permission android:name="android.permission.WAKE_LOCK" />
    <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>`
to my manifest (found them in the background_locator example), I could use:
`var userPosition = await Geolocator().getCurrentPosition();`
in a background fashion using background_fetch: ^0.6.0.

So it's only every 15 minutes but worked pretty well.
Will try background_locator now

**Comment by bhanuka96** [comment]
> Once I add:
> ` <uses-permission android:name="android.permission.INTERNET" /> <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" /> <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" /> <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" /> <uses-permission android:name="android.permission.WAKE_LOCK" /> <uses-permission android:name="android.permission.FOREGROUND_SERVICE" /> <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>`
> to my manifest (found them in the background_locator example), I could use:
> `var userPosition = await Geolocator().getCurrentPosition();`
> in a background fashion using background_fetch: ^0.6.0.
> 
> So it's only every 15 minutes but worked pretty well.
> Will try background_locator now

I faced this issue on background_locator https://github.com/rekab-app/background_locator/issues/167

And then I moved to the this package https://github.com/transistorsoft/flutter_background_geolocation this package is great

**Comment by smolugu** [comment]
@Dereumaux-Adrien thank you for the update! Really appreciate it. 15 minutes is good enough for me . Will try it..

**Comment by Dereumaux-Adrien** [comment]
@bhanuka96 Indeed, it works flawlessly but as I don't yet know if my project has any commercial value I don't intend to spend money on plugins.
I already value transistorsoft as background_fetch is from them as well.
If the project manages to yield a good result, I guess I'll switch to flutter_background_geolocation later on.

**Comment by Dereumaux-Adrien** [comment]
TLDR for people new to this thread

If you can pay, https://pub.dev/packages/flutter_background_geolocation works easily and well.
If you can spend time on it, you should manage to make https://pub.dev/packages/background_locator work (couldn't myself yet).
If you want to begin fast, can't spend money yet and are okay with a background check happening every 15~ minutes : https://pub.dev/packages/background_fetch with https://pub.dev/packages/geolocator with manifest changes written [here](https://github.com/Baseflow/flutter-geolocator/issues/53#issuecomment-748626471)

**Comment by christocracy** [comment]
@Dereumaux-Adrien email me for a free license.

**Comment by wrabit** [comment]
You can also use WorkManager as detailed here: https://medium.com/@pierre.sabot/how-to-fetch-user-location-in-background-with-flutter-e3494021bdf5

**Comment by temala** [comment]
I used https://pub.dev/packages/background_location to track user location in background and save to firestore

**Comment by zeldan** [comment]
You can try use location package, it has support for background (for 2 months) - https://pub.dev/packages/location
See Readme for instructions - https://github.com/Lyokone/flutterlocation

**Comment by steinmetz** [comment]
@zeldan it doesn't keep working in background if you kill the app, right? As far as I tested the example app, it stops listening in background when you kill the app.

**Comment by mvanbeusekom** [comment]
Also I am wondering if you don't really would like a callback to be called when a location arrives in the background (so you could proces the location, or safe it somewhere so it can be processed when the app comes back in the foreground). 

Not sure how this is handled at the moment in the `location` plugin. But that is how I was thinking to go about it for the `geolocator`. 

By the way it is not really respectful referring to other plugins.

**Comment by PaoloP98** [comment]
> You can also use WorkManager as detailed here: https://medium.com/@pierre.sabot/how-to-fetch-user-location-in-background-with-flutter-e3494021bdf5
> 
> It still limits to 15 minutes for Android, but it's another approach, simple API.

I tried to get the position periodically every 15 minutes and everything works well.
But by setting a periodic event every hour I cannot get the position. The workmanager works (the notification appear) but I can't get the location.
Have you tried to set a periodic event? If so, how long?

**Comment by mbilalakmal** [comment]
@mvanbeusekom should I open a seperate issue for supporting location updates while the app is in the foreground?
By foreground, I mean the application is running but the user has tapped the home button, switched to a different app, or the screen has been turned off. For Android, we need to add a foreground service that displays a persistent notification alerting the user that the app is using their location. For iOS, a green icon is displayed in the statusbar (this is allowed under NSLocationWhenInUsage permission).
I think supporting this would be much easier and solve many use cases where actual background location updates is not required. (Background location access means the app is launched if terminated when location updates are fetched and the background service can run independent of the app).

**Comment by guigzp** [comment]
Any updates on this? I just wanted to listen to updates while the app is not killed, like @mbilalakmal said. The [location package](https://pub.dev/packages/location) already does this, but I prefer this package structure and API.

**Comment by jesussmile** [comment]
> I used https://pub.dev/packages/background_location to track user location in background and save to firestore

Hi @temala I'm trying to do exactly what you have done! Could you please help me out with a reference or code snippet or anything to use background location with firebase ?

**Comment by turkiasoft** [comment]
@nanjadev  thanks you...

**Comment by ICP-Theo** [comment]
I have a bike app that get location every second. Up to version 7.03 we could get the location when the app was running in the backgound ( background as defined by mbilalakmal, see his comment 14 mar 2021). After upgrading we no longer could get the location when the app was in the backgound and we were forced to downgrade to version 7.03.
I just tested the latest version (version 8.01) but also this version prevents us to get the location with the app in the background.
Perhaps you can see what the difference is coming from version 7.03 and solve the issue.

**Comment by ICP-Theo** [comment]
Is it possible that one of the dependencies is causing the background permission problems?
Anyone any suggestions on this problem??

**Comment by cto-leaps** [comment]
> I have a bike app that get location every second. Up to version 7.03 we could get the location when the app was running in the backgound ( background as defined by mbilalakmal, see his comment 14 mar 2021). After upgrading we no longer could get the location when the app was in the backgound and we were forced to downgrade to version 7.03.
> I just tested the latest version (version 8.01) but also this version prevents us to get the location with the app in the background.
> Perhaps you can see what the difference is coming from version 7.03 and solve the issue.

@ICP-Theo Do you have example of code of using a Foreground Service and Geolocator (7.0.3)? I've been trying to implement something to no avaiL...

**Comment by ICP-Theo** [comment]
@cto-leaps 
My focus was IOS to start with, so I set all the keys in the info.plist:
<key>UIBackgroundModes</key>
	<array>
		<string>fetch</string>
		<string>location</string>
		<string>bluetooth-central</string>
	</array>
	<key>NSLocationAlwaysAndWhenInUseUsageDescription</key>
	<string>We need your location to create a route that you have cycled. Yes even when you put your devices in sleep mode you want to track your route</string>
	<key>NSLocationAlwaysUsageDescription</key>
	<string>We need your location to create a route that you have cycled. Yes even when you put your devices in sleep mode you want to track your route</string>
	<key>NSLocationWhenInUseUsageDescription</key>
	<string>We need your location to create a route that you have cycled. Yes even when you put your devices in sleep mode you want to track your route</string>

In this version I did not pay any attention to Android but I had an older version written in C++ with a foreground service to get the location when the app ran in the background and that worked quite well.
When I have solved this issue for IOS I wil look into the foreground service for Android

**Comment by jesussmile** [comment]
> 

exactly the same problem, did you find any fix ?

**Comment by jesussmile** [comment]
> I have a bike app that get location every second. Up to version 7.03 we could get the location when the app was running in the backgound ( background as defined by mbilalakmal, see his comment 14 mar 2021). After upgrading we no longer could get the location when the app was in the backgound and we were forced to downgrade to version 7.03. I just tested the latest version (version 8.01) but also this version prevents us to get the location with the app in the background. Perhaps you can see what the difference is coming from version 7.03 and solve the issue.

Exactly the same issue, did you find any fix ?

**Comment by cto-leaps** [comment]
> Exactly the same issue, did you find any fix ?

Context: My fitness app is used for my user to track their position while they perform certain activities they launch themselves. I am not trying to detect or track them when they are not using the app. 
I do, however, need to track them when the app is not in the foreground but an activity they have launched is still running.  

In the end, I used two separate approaches. 

- On iOS, I simply used GeoLocator and added the "Background updates" permissions in the plist. I ask the user for "While the app is in use" permission. I subscribe to the Stream from the main isolate and I get position updates every x meters whether the app is in foreground or background. It works flawlessly. No need for an "Always" permission as I don't need updates when the app is killed. 

- Different story on Android: when the user launches an activity, I launch a foreground service. It's that foreground service who's in charge of listening to the GeoLocator stream and sends the updates to the main isolate, keep the app "alive" even if it's in the background. 
There again, I only need the "While in use" permission. I kill the foreground when the activity is stopped by the user. 
I initially try to spin the foreground service only when the app was going in the background, but the GeoLocator was getting lost with the streams...

I use the latest version of GeoLocator and everything seems to work fine on both platforms. I'm still in test on Android and waiting for the app to be approved. But so far, so good. 

I hope this helps, @jesussmile !

**Comment by ICP-Theo** [comment]
@jesussmile 
Like I said my focus was iOS, so I started my own solution to get the location to see what was going on in the background.
Then xCode mentioned I should increase the target deployment version, so I changed the 'global platform for iOS' in the podfile from 9.0 to 14.1. After that I could get the location even in the background. I went back to Geolocator (global platform for iOS still 14.1) and also now I get the location when the app runs in the background. I know that from version 7.1.0 they changed to a federated plugin and maybe settings for older target versions were not the same (just a guess).
Anyway, this solved it for me for iOS. Will be checking Android soon but priorities are elsewhere at the moment.

I hope this helps someone

**Comment by kakashysen** [comment]
I don't know if background location tracking means that the app should stay tracking the user location even if it's killed, in my case I need to track the location when the app goes to background, I had some issues with that like the `listener` is not executed however I notice one issue in my android configuration I was forcing to use the LocationManager instead of the FusedLocationProviderClient, I just set this configuration `forceLocationManager: false,` and the background location works 

```
locationSettings = AndroidSettings(
          accuracy: LocationAccuracy.high,
          distanceFilter: 100,
          forceLocationManager: false,
          intervalDuration: const Duration(seconds: 10),
          foregroundNotificationConfig: const ForegroundNotificationConfig(
            notificationText:
                "Example app will continue to receive your location even when you aren't using it",
            notificationTitle: "Running in Background",
            enableWakeLock: true,
          ));

_positionStream =
        Geolocator.getPositionStream(locationSettings: locationSettings)
            .listen((Position position) {
      _lastPosition = _currentPosition;
      _currentPosition = position;
...
}
```

**Comment by spehj** [comment]
Could you let me know if you plan to support background location updates when the app is terminated?
